### ISSUE TITLE
Unable to detect if unauthenticated right after redirect login

### ISSUE THREAD

**Issue by jasonjohnlee** [AzureAD/microsoft-authentication-library-for-js#2360]
## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [X] `@azure/msal-browser@2.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Description
Hi, I have a React.js app and I want to ensure that only authenticated users can access certain pages/routes.  We login using the `loginRedirect` and when MS redirects back to our application I understand we should await `handleRedirectPromise` to determine if the user has authenticated correctly.  The problem is that the first return value from `handleRedirectPromise` is null and so our application assumes that the user is not authenticated.  However, a subsequent invocation of `handleRedirectPromise` returns with an authenticated response as expected.  Am I doing something wrong?  What is the correct pattern to determine if the user is **unauthenticated**?  

I did see the [page load example](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-browser-samples/VanillaJSTestApp2.0/app/onPageLoad/auth.js) and tried to model my code after that.  Other issues like [this one](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/2160) are trying to determine if the user is authenticated, not unauthenticated.

Here's an excerpt of my authentication provider code that wraps MSAL as a React Context.  My React route handler calls the `isAuthenticated` function to determine if the user is authenticated.  The problem is that the `handleRedirectPromise` in the `isAuthenticated` function returns null so the app thinks the user is unauthenticated and redirects the user back to the login page, but after that function that does the authentication check completes, the `handleRedirectPromise` promise completes in my `setMsalInstance` method with an authenticated response...but it's too late, the app has already assumed the user is not authenticated.

```
export const AuthenticationProvider: Function = ({
  config,
  children,
}: AuthenticationContextProps) => {

  (function setMsalInstance() {
    if (!msalInstance) {
      ...
      msalInstance = new msal.PublicClientApplication(msalConfig);
      
      msalInstance.handleRedirectPromise()
        .then((tokenResponse) => {
            //After my isAuthenticated method completes, this returns with a tokenResponse showing user is authenticated.
        })
        .catch((error) => {throw error});
    }
  })();

  const isAuthenticated = async (): Promise<boolean> => {
    const tokenResponse = await msalInstance.handleRedirectPromise();

    if (tokenResponse != null) {
      return true;
    }
    else{
      return msalInstance.getAllAccounts().length > 0; //<---tokenResponse is ALWAYS null right after a redirect from MS, and there are no accounts.
    };
  }
```

**Comment by tnorling** [comment]
@jasonjohnlee `handleRedirectPromise` will only resolve once. What seems to be happening here is that you're calling it twice and the 2nd one (inside isAuthenticated) is immediately returning null because the first has already started doing what it needs to do. You can kind of think of `handleRedirectPromise` as the resolution of the unresolved promise from `loginRedirect`.

What you should do instead is ensure that you wait for the first `handleRedirectPromise` to resolve before calling any other msal function. Then in your `isAuthenticated` function all you need to do is call `getAllAccounts`

**Comment by jasonjohnlee** [comment]
Thanks @tnorling !  Removing the `handleRedirectPromise` in `isAuthenticated` solved the problem of the null response.  It also required me to redesign/rethink how I performed this unauthenticated check.  I originally thought I'd be able to do a simple if statement in the router to see if the user was authenticated before going to certain routes, but I had to change my approach to make my `AuthenticationProvider` a React pipeline Component in addition to a context.  The component will return a null rendering if the user is unauthenticated (which the user is initially), then will re-render with the requested route(which is a child component of the `AuthenticationProvider`) when `handleRedirectPromise` returns with a response or there are logged in accounts.  If `handleRedirectPromise` returns and there is no response or logged in accounts, I redirect to the sign in page.
